cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java

```diff
@@ -390,11 +347,89 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
                                                     "is created.", difference, namespaceId));
     }
     nsStore.update(builder.build());
+    // refresh the cache with new meta
+    namespaceMetaCache.refresh(namespaceId.toEntityId());
+  }
+
+  /**
+   * Lists all namespaces
+   *
+   * @return a list of {@link NamespaceMeta} for all namespaces
+   */
+  @Override
+  public List<NamespaceMeta> list() throws Exception {
+    List<NamespaceMeta> namespaces = nsStore.list();
+    Principal principal = authenticationContext.getPrincipal();
+    final co.cask.cdap.api.Predicate<EntityId> filter = authorizationEnforcer.createFilter(principal);
+    return Lists.newArrayList(
+      Iterables.filter(namespaces, new com.google.common.base.Predicate<NamespaceMeta>() {
+        @Override
+        public boolean apply(NamespaceMeta namespaceMeta) {
+          return filter.apply(namespaceMeta.getNamespaceId());
+        }
+      })
+    );
+  }
+
+  /**
+   * Gets details of a namespace
+   *
+   * @param namespaceId the {@link Id.Namespace} of the requested namespace
+   * @return the {@link NamespaceMeta} of the requested namespace
+   * @throws NamespaceNotFoundException if the requested namespace is not found
+   * @throws UnauthorizedException if the namespace is not authorized to the logged-user
+   */
+  @Override
+  public NamespaceMeta get(Id.Namespace namespaceId) throws Exception {
+    try {
+      return namespaceMetaCache.get(namespaceId.toEntityId());
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof NamespaceNotFoundException || cause instanceof IOException ||
+        cause instanceof UnauthorizedException) {
+        throw (Exception) cause;
+      }
+      throw e;
+    }
+  }
+
+  /**
+   * Checks if the specified namespace exists
+   *
+   * @param namespaceId the {@link Id.Namespace} to check for existence
+   * @return true, if the specifed namespace exists, false otherwise
+   */
+  @Override
+  public boolean exists(Id.Namespace namespaceId) throws Exception {
+    try {
+      get(namespaceId);
+      return true;
+    } catch (NamespaceNotFoundException e) {
+      return false;
+    }
+  }
+
+  @VisibleForTesting
+  Map<NamespaceId, NamespaceMeta> getCache() {
+    return namespaceMetaCache.asMap();
+  }
+
+  private NamespaceMeta fetchNamespaceMeta(NamespaceId namespaceId) throws Exception {
+    NamespaceMeta ns = nsStore.get(namespaceId.toId());
+    if (ns == null) {
+      throw new NamespaceNotFoundException(namespaceId.toId());
+    }
+    Principal principal = authenticationContext.getPrincipal();
```

Comment:
please move this out of the cache and into the ``get`` method. Also, ``exists()`` should not enforce authorization